---
title: "01_run_DESEQ2"
author: "JR"
date: "2024-04-19"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(IRanges)
library(dplyr)
library(tidyr)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
```

# Goal to run DESEQ2 and get familiar with results
# Please note that we need to organize our new directory with "analysis, figures and results directories"
Using DESEQ2 we will determine which genes change in RNA expression levels
at different times of dox exposure. Let's officially start doing some analysis!

We are going to start slowly and stepwise to get a feel for DESEQ2.
It is highly recommended that you read these documents for deeper dives!

https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html


# Let's start by loading in our counts_matrix and sample info
# IMPORTANT: careful of order when loading Rdata :)
```{r loading in counts and sample info}
load("../05_RNAseq_in_R_studio/results/deseq_samples.RData", verbose = T)
load("../05_RNAseq_in_R_studio/results/count_files.RData", verbose = T)
```

#DESeq: Tests for Differential Expression of Genes in RNASeq Data based on a negative binomial model

# 1) Prepare for DESeq2 by creating a dds (DESEQ Data Structure) 

DESeqDataSetFromMatrix = Create a DESeq Data Set from Matrix w these parameters
  countData = selects count data to read in  // counts_filtered
      ** DESeq requires all the integers to be rounded ** 
  colData = what env variable to use for the sample sheet // deseq_samples
  design = _(what does ~ do)_ condition to factor   //timepoint
  
```{r run DEseqDataSetFromMatrix }
dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              colData = deseq_samples,
                              design = ~ time_point)
```

# 2) Run DESeq
run DESeq on dds data frame and save to dds

```{r run DEseqDataSetFromMatrix }
dds <- DESeq(dds)
```

# 3) Differential Expression Results

resultsNames = retrieves names of results
  dds = where from
[output] = what analyses DESeq2 Performed

```{r results}
resultsNames(dds) # resultsNames: retrieves names of results


# Nice we see this output:
# "Intercept" "time_point_12_vs_0" "time_point_24_vs_0" "time_point_48_vs_0" "time_point_96_vs_0"
# DESEQ found genes differential expression values for each time point relative to 0 time point.
```

# RESULT: we have compared gene expression values relative to 0 timepoint.
It is good to start making a notation of results you wan to keep track of.
In this case it is key to record what analysis was done.

# Next:Running DeSeq to create rlog normalzied counts!
Although a lot of normalization has already occurred from
Salmon (where we got the counts) we now need to determine the 
distribution of expression values and normalize again to prevent
low abundance genes inflating significant differential expression. 


#  Normalize counts (rlog function)
This basically is rank counts normalized to std error in replicates.

```{r rlog }
rlog_counts <- rlog(dds, blind = TRUE)
```

# now we retrieve the values using the "assay" function that converts to rlog_counts
```{r}
rlog_counts_matrix <- assay(rlog_counts)

```

# Finally, let's save this rlog matrix to use in our future analyses and plotting.
We will save this as R Data Structure .rds -- this will keep the object stored properly
to be loaded into the environment to use for downstream analyses.

```{r}

write_rds(rlog_counts_matrix, "results/rlog_counts_mouse_dox_long.rds")

```

# Nice you now have run DESEQ2! 
There is a lot more to explore
In the meantime enjoy all the analyses DESEQ2 just did for you !